src/Engine/Data/EnvironmentTexture.cpp

```diff
@@ -0,0 +1,739 @@
+                            { { 0, -1, 0 }, { -1, 0, 0 }, { 0, 0, -1 } } };
+    auto sphericalPhi   = []( const Vector3& d ) {
+        Scalar p = std::atan2( d.x(), d.y() );
+        return ( p < 0 ) ? ( p + 2 * M_PI ) : p;
+    };
+    auto sphericalTheta = []( const Vector3& d ) { return std::acos( d.z() ); };
+#pragma omp parallel for
+    for ( int imgIdx = 0; imgIdx < 6; ++imgIdx ) {
+        // Alllocate the images
+        m_skyData[imgIdx] = new float[textureSize * textureSize * 4];
+        // Fill in pixels
+        for ( Scalar u = -1_ra; u < 1_ra; u += 2_ra / textureSize ) {
+            for ( Scalar v = -1_ra; v < 1_ra; v += 2_ra / textureSize ) {
+                Vector3 d = bases[imgIdx][0] + u * bases[imgIdx][1] + v * bases[imgIdx][2];
+                d         = d.normalized();
+                Vector2 st { w * sphericalPhi( d ) / ( 2 * M_PI ), h * sphericalTheta( d ) / M_PI };
+                // TODO : use st to access and filter the original envmap
+                // for now, no filtering is done. (eq to GL_NEAREST)
+                int s  = int( st.x() );
+                int t  = int( st.y() );
+                int cu = int( ( u / 2 + 0.5 ) * textureSize );
+                int cv = int( ( v / 2 + 0.5 ) * textureSize );
+
+                m_skyData[imgIdx][4 * ( cv * textureSize + cu ) + 0] =
+                    latlonPix[4 * ( t * w + s ) + 0];
+                m_skyData[imgIdx][4 * ( cv * textureSize + cu ) + 1] =
+                    latlonPix[4 * ( t * w + s ) + 1];
+                m_skyData[imgIdx][4 * ( cv * textureSize + cu ) + 2] =
+                    latlonPix[4 * ( t * w + s ) + 2];
+                m_skyData[imgIdx][4 * ( cv * textureSize + cu ) + 3] = 1;
+            }
+        }
+        flip_horizontally( m_skyData[imgIdx], textureSize, textureSize, 4 );
+    }
+    free( latlonPix );
+    m_width = m_height = textureSize;
+}
+
+void EnvironmentTexture::computeSHMatrices() {
+    for ( int i = 0; i < 9; i++ ) {
+        for ( int j = 0; j < 3; j++ ) {
+            m_shcoefs[i][j] = 0.f;
+        }
+    }
+    /// @todo replace this integration to use a sphere sampler ...
+    /// Must evaluate the elementary solid angle for each sample
+    const float dtheta = 0.005;
+    const float dphi   = 0.005;
+    for ( float theta = 0.f; theta < M_PI; theta += dtheta ) {
+        for ( float phi = 0.f; phi < 2.f * M_PI; phi += dphi ) {
+            auto x          = std::sin( theta ) * std::cos( phi );
+            auto y          = std::sin( theta ) * std::sin( phi );
+            auto z          = std::cos( theta );
+            float* thePixel = getPixel( x, y, z );
+            updateCoeffs( thePixel, -x, y, z, std::sin( theta ) * dtheta * dphi );
+        }
+    }
+    tomatrix();
+}
+
+void EnvironmentTexture::updateCoeffs( float* hdr, float x, float y, float z, float domega ) {
+    /******************************************************************
```

Comment:
Could it make sense to move the following documentation to the public header ?

Reply:
I don't think so, this is clearly only a private implementation that might evolve. I added in the public header the reference to the paper the actual implementation is based on and a brief doc on this method.